### ISSUE TITLE
CA1859 recommends change that fails due to inconsistent accessibility

### ISSUE THREAD

**Issue by bgrainger** [dotnet/roslyn-analyzers#6659]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859): Change return type of method M from X to Y for improved performance.

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: 8.0.100-preview.4.23260.5

_and also_

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: [8.0.0-preview1.23165.1](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers/8.0.0-preview1.23165.1) (Latest preview)

### Describe the bug

CA1859 is triggered on code that returns a nested type. The diagnostic message is `Change return type of method 'Method' from 'System.IDisposable' to 'NestedReturnType.Program.NestedClass1.NestedClass2' for improved performance`. If the change is made, the updated code fails to compile with the error `error CS0050: Inconsistent accessibility: return type 'Program.NestedClass1.NestedClass2' is less accessible than method 'Program.NestedClass1.Method()'`.

### Steps To Reproduce

Compile the following code.

<details>
  <summary>CSPROJ file</summary>

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
    <AnalysisMode>AllEnabledByDefault</AnalysisMode>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.CodeAnalysis.NetAnalyzers" Version="8.0.0-preview1.23165.1">
      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

</Project>
```
</details>

```csharp
namespace NestedReturnType;

public class Program
{
	public static void Main(string[] args)
	{
	}

	private static class NestedClass1
	{
		public static IDisposable Method()
		{
			return new NestedClass2();
		}

		private sealed class NestedClass2 : IDisposable
		{
			public void Dispose()
			{
			}
		}
	}
}
```

### Expected behavior

No CA1859 warning.

### Actual behavior

CA1859 warning that recommends an invalid change.

## Additional context

Possibly a duplicate of https://github.com/dotnet/roslyn-analyzers/issues/6479, although .NET SDK 8 Preview 4 and the latest NuGet package have both been released since then, so I expected them to contain the fix.
